cocos/audio/android/AudioDecoder.cpp

```diff
@@ -0,0 +1,666 @@
+    /* Callback for "prefetch" events, here used to detect audio resource opening errors */
+    static void prefetchEventCallback( SLPrefetchStatusItf caller,  void* context, SLuint32 event)
+    {
+        AudioDecoder* thiz = reinterpret_cast<AudioDecoder*>(context);
+        thiz->prefetchCallback(caller, event);
+    }
+
+    static void decPlayCallback(SLAndroidSimpleBufferQueueItf queueItf, void* context)
+    {
+        AudioDecoder* thiz = reinterpret_cast<AudioDecoder*>(context);
+        thiz->decodeToPcmCallback(queueItf);
+    }
+
+    static void decProgressCallback(SLPlayItf caller, void* context, SLuint32 event)
+    {
+        AudioDecoder* thiz = reinterpret_cast<AudioDecoder*>(context);
+        thiz->decodeProgressCallback(caller, event);
+    }
+};
+
+AudioDecoder::AudioDecoder(SLEngineItf engineItf, const std::string &url, int sampleRate)
+        : _engineItf(engineItf)
+        , _url(url)
+        , _playObj(nullptr)
+        , _pcmMetaData(nullptr)
+        , _formatQueried(false)
+        , _prefetchError(false)
+        , _counter(0)
+        , _numChannelsKeyIndex(-1)
+        , _sampleRateKeyIndex(-1)
+        , _bitsPerSampleKeyIndex(-1)
+        , _containerSizeKeyIndex(-1)
+        , _channelMaskKeyIndex(-1)
+        , _endiannessKeyIndex(-1)
+        , _eos(false)
+        , _sampleRate(sampleRate)
+        , _assetFd(0)
+{
+    memset(_pcmData, 0, sizeof(_pcmData));
+    auto pcmBuffer = std::make_shared<std::vector<char>>();
+    pcmBuffer->reserve(4096);
+    _result.pcmBuffer = pcmBuffer;
+}
+
+AudioDecoder::~AudioDecoder()
+{
+    if (_pcmMetaData != nullptr)
+    {
+        free(_pcmMetaData);
+        _pcmMetaData = nullptr;
+    }
+
+    if (_assetFd > 0)
+    {
+        ::close(_assetFd);
+        _assetFd = 0;
+    }
+
+    SL_DESTROY_OBJ(_playObj);
+}
+
+bool AudioDecoder::start(const FdGetterCallback& fdGetterCallback)
+{
+    SLresult  result;
+
+    /* Objects this application uses: one audio player */
+    SLObjectItf  player;
+
+    /* Interfaces for the audio player */
+    SLAndroidSimpleBufferQueueItf decBuffQueueItf;
+    SLPrefetchStatusItf           prefetchItf;
+    SLPlayItf                     playItf;
+    SLMetadataExtractionItf       mdExtrItf;
+
+    /* Source of audio data for the decoding */
+    SLDataSource      decSource;
+    SLDataLocator_URI decUri;
+    
+    /* Data sink for decoded audio */
+    SLDataSink                decDest;
+    SLDataLocator_AndroidSimpleBufferQueue decBuffQueue;
+    SLDataFormat_PCM          pcm;
+
+    SLboolean required[NUM_EXPLICIT_INTERFACES_FOR_PLAYER];
+    SLInterfaceID iidArray[NUM_EXPLICIT_INTERFACES_FOR_PLAYER];
+
+    /* Initialize arrays required[] and iidArray[] */
+    for (int i=0 ; i < NUM_EXPLICIT_INTERFACES_FOR_PLAYER ; i++) {
+        required[i] = SL_BOOLEAN_FALSE;
+        iidArray[i] = SL_IID_NULL;
+    }
+
+    /* allocate memory to receive the PCM format metadata */
+    if (!_pcmMetaData) {
```

Comment:
Good point. 👍 
This code are most from Google sample code.  :)
Since `AudioDecoder::start` should only be invoked once by one instance, i think we have to check some variables more strictly.